external/solr/src/test/java/org/apache/stormcrawler/solr/persistence/SpoutTest.java

```diff
@@ -0,0 +1,169 @@
+import org.apache.storm.tuple.Tuple;
+import org.apache.stormcrawler.Metadata;
+import org.apache.stormcrawler.TestOutputCollector;
+import org.apache.stormcrawler.TestUtil;
+import org.apache.stormcrawler.persistence.Status;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Verifies that when having 2 Solr shards, documents will be processed by the 2 respective Spouts
+ */
+public class SpoutTest extends SolrContainerTest {
+
+    private StatusUpdaterBolt bolt;
+    private SolrSpout spoutOne;
+    private SolrSpout spoutTwo;
+    private TestOutputCollector boltOutput;
+    private TestOutputCollector spoutOneOutput;
+    private TestOutputCollector spoutTwoOutput;
+
+    private static final Logger LOG = LoggerFactory.getLogger(StatusBoltTest.class);
+
+    @Before
+    public void setup() throws IOException, InterruptedException {
+        container.start();
+        createCollection("status", 2);
+
+        bolt = new StatusUpdaterBolt();
+        boltOutput = new TestOutputCollector();
+
+        Map<String, Object> conf = new HashMap<>();
+        conf.put("scheduler.class", "org.apache.stormcrawler.persistence.DefaultScheduler");
+        conf.put("status.updater.cache.spec", "maximumSize=10000,expireAfterAccess=1h");
+        conf.put("solr.status.url", getSolrBaseUrl() + "/status");
+
+        conf.put("solr.status.routing.shards", 2);
+        conf.put(
+                "urlbuffer.class", "org.apache.stormcrawler.persistence.urlbuffer.SimpleURLBuffer");
+
+        bolt.prepare(conf, TestUtil.getMockedTopologyContext(), new OutputCollector(boltOutput));
+
+        spoutOne = new SolrSpout();
+        spoutTwo = new SolrSpout();
+        spoutOneOutput = new TestOutputCollector();
+        spoutTwoOutput = new TestOutputCollector();
+
+        spoutOne.open(conf, getContextForTask(0), new SpoutOutputCollector(spoutOneOutput));
+        spoutTwo.open(conf, getContextForTask(1), new SpoutOutputCollector(spoutTwoOutput));
+    }
+
+    @After
+    public void close() {
+        LOG.info("Closing updater bolt and SOLR container");
+        bolt.cleanup();
+        spoutOne.close();
+        spoutTwo.close();
+        container.close();
+        boltOutput = null;
+        spoutOneOutput = null;
+        spoutTwoOutput = null;
+    }
+
+    private TopologyContext getContextForTask(int taskId) {
+        Map<Integer, String> taskToComponent = new HashMap<>();
+        Map<String, List<Integer>> componentToTasks = new HashMap<>();
+
+        taskToComponent.put(0, "solrSpout");
+        taskToComponent.put(1, "solrSpout");
+        componentToTasks.put("solrSpout", Arrays.asList(0, 1));
+
+        return new TopologyContext(
+                null,
+                Map.of("storm.cluster.mode", "local"),
+                taskToComponent,
+                componentToTasks,
+                new HashMap<>(),
+                null,
+                null,
```

Comment:
When writing the test with the 2 spouts, I manually created this Storm `TopologyContext` object with most of the parameters set to `null`. Is this ok? Should we set anything else?

Reply:
no idea to be honest but it feels more complicated than what we've had to do for the other tests.
What about reusing ` TestUtil.getMockedTopologyContext() `?

Reply:
This didn't work, since what I wanted to test involved the SolrSpout calling [context.getComponentTasks()](https://github.com/apache/incubator-stormcrawler/blob/main/external/solr/src/main/java/org/apache/stormcrawler/solr/persistence/SolrSpout.java#L78) which in turn reads the [componentToTasks](https://github.com/apache/storm/blob/master/storm-client/src/jvm/org/apache/storm/task/GeneralTopologyContext.java#L105) for example. I started from the [FileSpoutTopologyContextMock](https://github.com/apache/incubator-stormcrawler/blob/main/core/src/test/java/org/apache/stormcrawler/spout/mocks/FileSpoutTopologyContextMock.java#L28-L48) and we could in principle have something similar if we want to have more such SolrSpout tests in the future.